src/Hub.cpp

```diff
@@ -252,6 +253,28 @@ void Hub::AssignVespeneHarvester(const sc2::Unit& refinery_) {
     worker->GatherVespene(refinery_);
 }
 
+void Hub::DistrubuteMineralWorker(const sc2::Unit* unit_) {
+    auto patches = gAPI->observer().GetUnits(sc2::IsVisibleMineralPatch(), sc2::Unit::Alliance::Neutral);
+    auto townhalls = gAPI->observer().GetUnits(IsNonSaturatedTownHall(), sc2::Unit::Alliance::Self);
+
+    if (townhalls.Empty())  // no non-Saturated bases, use any finished
+        townhalls = gAPI->observer().GetUnits(IsBuiltTownHall(), sc2::Unit::Alliance::Self);
+
+    const sc2::Unit* townhall_target = townhalls.GetClosestUnit(unit_->pos);
+
+    const sc2::Unit* mineral_target;
+    if (!townhall_target) {  // if no TownHalls existing (or landed?)
+        mineral_target = patches.GetClosestUnit(gAPI->observer().StartingLocation());
```

Comment:
Technically we have all required positions in the `Expansions` class. Is it possible to use/extend it somehow?

Reply:
Expansion could include a Tag for the current TownHall occupying an Expansion (SELF or ENEMY), that would fix the issue with knowing whether a TownHall is a macro or tied to an expansion. But I would include that with the Macro fix (#42).  Also, in about 30% of my test games, the Expansion-finding is not recognizing the natural or the natural+third expansions, so would have to fix that also before relying on Expansions.

Reply:
Although, I can't get the filter-chaining to work, so maybe I should just set the Tags on Expansions and loop through them for this, so I can avoid the filter-chaining for now.

Reply:
I intentionally added Expansions classes to use in situation like this and to mark corresponding regions as OWN, ENEMY, CONTESTED, although this logic doesn't fully implemented. Thus I suggest to give it a try.

Reply:
Changed this to use Expansions class.